src/clims/models/extensible.py

```diff
@@ -0,0 +1,184 @@
+
+
+class ExtensibleProperty(Model):
+    """
+    A class for properties that can be hooked up to `Extensible` objects.
+
+    Examples:
+
+    Already implemented:
+      * Bound to a `Substance`
+
+    Will be implemented:
+      * Bound to a `Container`
+      * Bound to a `Project`
+
+    Properties are versioned based on the version of the Extensible at the time they were
+    bound to it.
+
+    The type of a property is determined by a plugin when the system is upgraded. They are
+    modelled with `ExtensiblePropertyType`.
+    """
+    __core__ = True
+
+    version = models.IntegerField(null=False, default=1)
+
+    # True if this is the latest version of this instance
+    latest = models.BooleanField(null=False, default=True)
+
+    extensible_property_type = models.ForeignKey("clims.ExtensiblePropertyType")
+
+    extensible_property_value = models.ForeignKey("clims.ExtensiblePropertyValue")
+
+    def __init__(self, *args, **kwargs):
+        super(ExtensibleProperty, self).__init__(*args, **kwargs)
+        self.is_versioned = kwargs.get('is_versioned', True)
+
+    @property
+    def name(self):
+        if not self.extensible_property_type:
+            return None
+        return self.extensible_property_type.name
+
+    @property
+    def value(self):
+        if not self.extensible_property_value:
+            return None
+        raw_type = self.extensible_property_type.raw_type
+        if raw_type == ExtensiblePropertyType.INT:
+            return self.extensible_property_value.int_value
+        elif raw_type == ExtensiblePropertyType.FLOAT:
+            return self.extensible_property_value.float_value
+        elif raw_type == ExtensiblePropertyType.STRING:
+            return self.extensible_property_value.string_value
+        elif raw_type == ExtensiblePropertyType.BOOL:
+            return self.extensible_property_value.bool_value
+        elif raw_type == ExtensiblePropertyType.JSON:
+            return json.loads(self.extensible_property_value.string_value)
+        else:
+            raise ValueError("Unexpected raw type {}".format(raw_type))
+
+    def create_extensible_property_value(self, value):
+        """
+        Creates a property value that fits with the type of this property
+        """
+        ret = ExtensiblePropertyValue()
+        raw_type = self.extensible_property_type.raw_type
+        if raw_type == ExtensiblePropertyType.INT:
+            ret.int_value = value
+        elif raw_type == ExtensiblePropertyType.FLOAT:
+            ret.float_value = value
+        elif raw_type == ExtensiblePropertyType.STRING:
+            ret.string_value = value
+        elif raw_type == ExtensiblePropertyType.BOOL:
+            ret.bool_value = value
+        elif raw_type == ExtensiblePropertyType.JSON:
+            ret.string_value = json.dumps(value)
+        else:
+            raise ValueError("Unexpected raw type {}".format(raw_type))
+        ret.save()
```

Comment:
Shouldnt saving to db be done in the substance service?

Reply:
Good catch with not saving here! I'll fix that, it works in this case because the caller has a transaction, but that might not always be the case.

Regarding validation, the REST layer should certainly validate the input there. Not sure the service layer should, that it should rather trust that callers do that. Still uncertain about that aspect, we should definitely have more discussion on that.